### PR TITLE
use chromium cookie settings

### PR DESCRIPTION
**Pull request by bridiver**
## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [ ] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [x] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [x] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: common/shield_exceptions.cc [hunk 1]
```diff
@@ -38,46 +38,6 @@ bool IsBlockedResource(const GURL& gurl) {
...
-  if (any_match) {
-    return true;
-  }
-
-  // 1st-party-dependent whitelist
-  static std::map<GURL, std::vector<URLPattern> > whitelist_patterns = {
-    {
-      GURL("https://www.sliver.tv/"),
-      std::vector<URLPattern>({URLPattern(URLPattern::SCHEME_ALL,
-            "https://*.thetatoken.org:8700/*")})
```
Comment (snyderp) [inline common/shield_exceptions.cc]: @bridiver This exception needs to be in place to unbreak a partner site (sliver.tv).  I don't see it added back in elsewhere in the PR, so just wanted to flag it and make sure its carried over

File: chromium_src/services/network/network_context.cc [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "services/network/network_context.h"
+
+GURL GetURLForCookieAccess(const net::URLRequest& request) {
+  if (!request.site_for_cookies().is_empty())
+    return request.site_for_cookies();
+
+  if (request.network_isolation_key().IsFullyPopulated())
```
Comment (iefremov) [inline chromium_src/services/network/network_context.cc]: we want only the top frame oringin, don't we? This way we get top frame + iframe, not sure if this is needed (https://cs.chromium.org/chromium/src/net/base/network_isolation_key.cc?sq=package:chromium&dr=CSs&g=0&l=47)

File: components/content_settings/core/common/content_settings_util.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/content_settings/core/common/content_settings_util.h"
+
+#include "brave/components/brave_shields/common/brave_shield_constants.h"
+#include "components/content_settings/core/common/content_settings_types.h"
+#include "extensions/common/url_pattern.h"
```
Comment (iefremov) [inline components/content_settings/core/common/content_settings_util.cc]: deps for this

File: components/content_settings/core/common/content_settings_util.cc [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+bool IsWhitelistedCookieException(const GURL& primary_url,
+                                  const GURL& secondary_url,
+                                  bool allow_google_auth) {
+  // Note that there's already an exception for TLD+1, so don't add those here.
+  // Check with the security team before adding exceptions.
+
+  // 1st-party-INdependent whitelist
+  std::vector<URLPattern> fpi_whitelist_patterns = {};
```
Comment (iefremov) [inline components/content_settings/core/common/content_settings_util.cc]: in the future we may want to get rid of this in favor of something simpler... Bringing here `extensions` doesn't look good

File: components/content_settings/core/common/content_settings_util.cc [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    return false;
+  }
+  std::vector<URLPattern> &exceptions = i->second;
+  return std::any_of(exceptions.begin(), exceptions.end(),
+      [&secondary_url](const URLPattern& pattern) {
+        return pattern.MatchesURL(secondary_url);
+      });
+}
+
+bool IsAllowContentSetting(const ContentSettingsForOneType& content_settings,
```
Comment (iefremov) [inline components/content_settings/core/common/content_settings_util.cc]: Everything here was just moved from other files?

File: components/brave_shields/browser/tracking_protection_helper.cc [hunk 6]
```diff
@@ -7,7 +7,7 @@
 
 #include "base/task/post_task.h"
 #include "brave/browser/brave_browser_process_impl.h"
-#include "brave/common/brave_switches.h"
+// #include "brave/common/brave_switches.h"
```
Comment (iefremov) [inline components/brave_shields/browser/tracking_protection_helper.cc]: erase

File: components/brave_shields/browser/tracking_protection_service.cc [hunk 7]
```diff
@@ -190,25 +186,15 @@ void TrackingProtectionService::UpdateFirstPartyStorageTrackers(
       base::flat_set<std::string>(std::move(storage_trackers));
 }
 
-#endif
-
-bool TrackingProtectionService::ShouldStoreState(BraveCookieSettings* settings,
-                                                 HostContentSettingsMap* map,
+#else  // !BUILDFLAG(BRAVE_STP_ENABLED)
+bool TrackingProtectionService::ShouldStoreState(HostContentSettingsMap* map,
```
Comment (iefremov) [inline components/brave_shields/browser/tracking_protection_service.cc]: we can save a couple of lines by keeping a single function header and putting `#ifdef`s inside

File: components/brave_shields/browser/brave_shields_util.h [hunk 8]
```diff
@@ -39,12 +41,14 @@ void SetBraveShieldsEnabled(Profile* profile, bool enable, const GURL& url);
 // reset to the default value
 void ResetBraveShieldsEnabled(Profile* profile, const GURL& url);
```
Comment (iefremov) [inline components/brave_shields/browser/brave_shields_util.h]: probably combining this bunch of functions into a class could make things simpler

File: common/shield_exceptions.cc (hunk 1 above)
Comment (bridiver) [reply common/shield_exceptions.cc]: must be a rebase issue, I'll move it to the new location

File: chromium_src/services/network/network_context.cc (hunk 2 above)
Comment (bridiver) [reply chromium_src/services/network/network_context.cc]:
```  // The origin of the top frame of the page making the request (where
  // applicable). Note that this is experimental and may not always be set.
  // DEPRECATED: This was introduced for the cache key and will be removed once
  // |network_isolation_key| is set for most cases.```

File: components/content_settings/core/common/content_settings_util.cc (hunk 4 above)
Comment (bridiver) [reply components/content_settings/core/common/content_settings_util.cc]: this was just moved from another location https://github.com/brave/brave-core/pull/3130/files/6e0deceddd23685bb249bfe98ec0b18a047b63d1#diff-26a6ea1048877a68443a743095591513

File: components/brave_shields/browser/tracking_protection_service.cc (hunk 7 above)
Comment (bridiver) [reply components/brave_shields/browser/tracking_protection_service.cc]: well, I think the buildflags need to be changed in general for `BRAVE_STP_ENABLED` because it's hard to follow what's going on. Probably move all real STP code into a separate `.cc` file and keep this method here

File: common/shield_exceptions.cc (hunk 1 above)
Comment (bridiver) [reply common/shield_exceptions.cc]: updated https://github.com/brave/brave-core/pull/3130/files#diff-da6fb1ceb9261944ea46907846888791R57

File: components/content_settings/core/common/content_settings_util.cc (hunk 4 above)
Comment (bridiver) [reply components/content_settings/core/common/content_settings_util.cc]: and I'm not sure that extensions is actually needed, checking now

File: components/content_settings/core/common/content_settings_util.cc (hunk 5 above)
Comment (bridiver) [reply components/content_settings/core/common/content_settings_util.cc]: yes

File: components/brave_shields/browser/brave_shields_util.h (hunk 8 above)
Comment (bridiver) [reply components/brave_shields/browser/brave_shields_util.h]: they would probably go in BraveShieldsService once we combine with the new BraveShieldsPrefService and convert to a BCKS. Until then I think we might as well leave them here

File: chromium_src/services/network/network_context.cc (hunk 2 above)
Comment (iefremov) [reply chromium_src/services/network/network_context.cc]: I'm talking about the top frame origin contained in `network_isolation_key`. Currently `network_isolation_key().ToString()` will return slightly more than just origin.

Comment (iefremov) [reply chromium_src/services/network/network_context.cc]: To be clear, I'm suggesting to use `request.network_isolation_key().GetTopFrameOrigin()`

File: components/content_settings/core/common/content_settings_util.cc [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      fpi_whitelist_patterns.end(),
+      [&secondary_url](const URLPattern& pattern) {
+        return pattern.MatchesURL(secondary_url);
+      });
+  if (any_match) {
+    return true;
+  }
+
+  // 1st-party-dependent whitelist
+  static std::map<GURL, std::vector<URLPattern> > whitelist_patterns = {
```
Comment (iefremov) [inline components/content_settings/core/common/content_settings_util.cc]: using `static` this way is deprecated, mind fixing it or posting a TODO/issue?

File: components/content_settings/core/browser/brave_content_settings_pref_provider.h [hunk 10]
```diff
@@ -35,10 +40,30 @@ class BravePrefProvider : public PrefProvider {
...
+  void OnContentSettingChanged(const ContentSettingsPattern& primary_pattern,
+                               const ContentSettingsPattern& secondary_pattern,
+                               ContentSettingsType content_type,
+                               const std::string& resource_identifier) override;
+  void NotifyChanges(const std::vector<Rule>& rules, const bool incognito);
 
   // PrefProvider::pref_change_registrar_ alreay has plugin type.
   PrefChangeRegistrar brave_pref_change_registrar_;
 
+  std::map<bool, std::vector<Rule>> cookie_rules_;
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.h]: I'd prefer to have a comment describing this mapping from boolean into something

File: components/content_settings/core/browser/brave_content_settings_pref_provider.h [hunk 11]
```diff
@@ -35,10 +40,30 @@ class BravePrefProvider : public PrefProvider {
...
+ private:
+  void UpdateCookieRules(ContentSettingsType content_type, bool incognito);
+  void OnCookieSettingsChanged(ContentSettingsType content_type);
+
+  // content_settings::Observer overrides:
+  void OnContentSettingChanged(const ContentSettingsPattern& primary_pattern,
+                               const ContentSettingsPattern& secondary_pattern,
+                               ContentSettingsType content_type,
+                               const std::string& resource_identifier) override;
+  void NotifyChanges(const std::vector<Rule>& rules, const bool incognito);
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.h]: const looks useless?

Comment (iefremov) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.h]: and this method doesn't override anything, so we can move it, or separate with a blank line

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 12]
```diff
@@ -31,13 +119,19 @@ BravePrefProvider::BravePrefProvider(PrefService* prefs,
               info->type(), prefs_, &brave_pref_change_registrar_,
               info->pref_name(),
               is_incognito_,
-              base::Bind(&PrefProvider::Notify, base::Unretained(this)))));
-      return;
+              base::Bind(&PrefProvider::Notify, weak_factory_.GetWeakPtr()))));
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: it is safe to keep `Unretained` here (since we own `content_settings_prefs_`), so weak factory is not needed

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 13]
```diff
@@ -9,15 +9,103 @@
...
 namespace content_settings {
 
+namespace {
+
+Rule CloneRule(const Rule& rule, bool reverse_patterns = false) {
+  auto secondary_pattern = rule.secondary_pattern;
+  if (secondary_pattern.Compare(
+          // TODO(bridiver) convert this to a brave_shields constant
+          ContentSettingsPattern::FromString("https://firstParty/*")) ==
+      ContentSettingsPattern::IDENTITY) {
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: do we really need to compare like this? Why not just compare with `==`?

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 14]
```diff
@@ -9,15 +9,103 @@
...
+    secondary_pattern = rule.primary_pattern;
+  }
+
+  // brave plugin rules incorrectly use the embedded url as the primary
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: maybe post an issue if we want to migrate this at some point...

**Comment by iefremov** [comment]
cookies
I've played around and found some inconsistencies between changing cookies settings normally via settings and changing via shields. Not sure if this is intended - see the gif (everything happens in incognito window, but that actually doesn't matter)

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 15]
```diff
@@ -57,9 +151,187 @@ bool BravePrefProvider::SetWebsiteSetting(
...
+           CONTENT_SETTINGS_TYPE_COOKIES,
+           "");
+  }
+}
+void BravePrefProvider::OnCookieSettingsChanged(ContentSettingsType content_type) {
+  UpdateCookieRules(content_type, true);
+  UpdateCookieRules(content_type, false);
+}
+
+void BravePrefProvider::OnContentSettingChanged(
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: Just to improve my understanding - what triggers it?

**Comment by iefremov** [comment]
> cookies
> I've played around and found some inconsistencies between changing cookies settings normally via settings and changing via shields. Not sure if this is intended - see the gif (everything happens in incognito window, but that actually doesn't matter)

also, `* embedded on https://ya.ru` stays even after I switch cookies shield back to "block cross-site cookies"

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 16]
```diff
@@ -57,9 +151,187 @@ bool BravePrefProvider::SetWebsiteSetting(
...
+  }
+
+  // Notify brave cookie changes as CONTENT_SETTINGS_TYPE_COOKIES
+  if (content_type == CONTENT_SETTINGS_TYPE_PLUGINS) {
+    // PostTask here to avoid content settings autolock DCHECK
+    base::PostTaskWithTraits(
+          FROM_HERE,
+          {content::BrowserThread::UI, base::TaskPriority::USER_VISIBLE},
+          base::BindOnce(&BravePrefProvider::NotifyChanges,
+                         weak_factory_.GetWeakPtr(),
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: maybe `base::ThreadTaskRunnerHandle::Get()->PostTask(FROM_HERE, base::BindOnce(&Task));` or something to avoid adding weak factory

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc [hunk 17]
```diff
@@ -57,9 +151,187 @@ bool BravePrefProvider::SetWebsiteSetting(
...
+                 new_rule.secondary_pattern == old_rule.secondary_pattern &&
+                 ValueToContentSetting(&new_rule.value) ==
+                    ValueToContentSetting(&old_rule.value);
+        });
+    if (match == old_rules.end()) {
+      brave_cookie_updates.push_back(CloneRule(new_rule));
+    }
+  }
+
+  // find any removed rules
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: nit - everywhere in this file CC wants comments to look like sentences: `Find any removed rules.`

File: components/content_settings/core/browser/brave_content_settings_pref_provider.h [hunk 18]
```diff
@@ -20,13 +24,14 @@ namespace content_settings {
 // Because of this reasion, shields configuration was also ephemeral.
 // However, we want shilelds configuration persisted. To do this, we make
 // EphemeralProvider ignore shields type and this class handles.
-class BravePrefProvider : public PrefProvider {
+class BravePrefProvider : public PrefProvider,
```
Comment (iefremov) [inline components/content_settings/core/browser/brave_content_settings_pref_provider.h]: So with cookies and shields, this is actually not a `Pref` provider, since we do not actually store stuff in prefs, but not sure what is the better place for it.

Comment (iefremov) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.h]: Maybe at least it's worth to expand comments regarding all the tricks

File: components/brave_shields/browser/tracking_protection_service.cc (hunk 7 above)
Comment (iefremov) [reply components/brave_shields/browser/tracking_protection_service.cc]: @jumde wdyt? maybe file an issue?

File: components/brave_shields/browser/brave_shields_util.cc [hunk 19]
```diff
@@ -97,6 +98,26 @@ ControlType ControlTypeFromString(const std::string& string) {
...
+// use AdBlockPrefService from
+// https://github.com/brave/brave-core/pull/2988/files#diff-37b733c3fa764e3f6d3723960f004952R15
+void SetCookiePrefDefaults(Profile* profile) {
+  auto type = GetCookieControlType(profile, GURL());
+  profile->GetPrefs()->SetBoolean(prefs::kBlockThirdPartyCookies,
+      type == ControlType::BLOCK_THIRD_PARTY);
+
+  if (type == ControlType::BLOCK) {
+    profile->GetPrefs()->SetInteger(
+        "profile.default_content_setting_values.cookies",
```
Comment (iefremov) [inline components/brave_shields/browser/brave_shields_util.cc]: nit: move it somewhere like `shield_pref_names.h`?

Comment (iefremov) [reply components/brave_shields/browser/brave_shields_util.cc]: ah, they are generic, nevermind

**Comment by iefremov** [comment]
https://github.com/brave/brave-browser/issues/2223 still reproduces with NetworkService and these changes :-/ That means we still have an overloaded IO thread

File: components/content_settings/core/browser/brave_content_settings_pref_provider.h (hunk 18 above)
Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.h]: not sure what you mean, it is a pref provider and that is where the content settings are stored. This is how all persistent content settings are retrieved

File: chromium_src/services/network/network_context.cc (hunk 2 above)
Comment (bridiver) [reply chromium_src/services/network/network_context.cc]: I see, I'll make that change

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc (hunk 13 above)
Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: in this case we can, I think this changed from a different comparison and just left the `Compare`

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc (hunk 12 above)
Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: yea, I think this is also leftover from a previous iteration. I'll remove

File: components/content_settings/core/browser/brave_content_settings_pref_provider.cc (hunk 15 above)
Comment (bridiver) [reply components/content_settings/core/browser/brave_content_settings_pref_provider.cc]: `ObservableProvider::NotifyObservers` - we register as an observer for that here https://github.com/brave/brave-core/pull/3130/files#diff-9981585c3d9184fc54b96697c59d2028R127

**Comment by bridiver** [comment]
@iefremov the behavior you're showing in the video is expected and intentional. We currently have no way to remove shields site exceptions (you can only toggle whether they are allow or block overrides) and there is also no way to distinguish between an override and a default setting for a site. The pattern shown `*` is correct for blocking all cookies embedded in ya.ru which is exactly what the chromium cookie setting shows (although it's partially obscured with ellipsis because the the window is too narrow to display the full text)
